opengrok-indexer/src/main/java/org/opengrok/indexer/util/HeadHandler.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
+ */
+
+package org.opengrok.indexer.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The purpose of this class is to provide StreamHandler that limits the output
+ * to specified number of lines. Compared to {@code SpoolHandler} it consumes
+ * limited amount of heap.
+ */
+public class HeadHandler implements Executor.StreamHandler {
+    private int numlines;
+
+    private final List<String> lines = new ArrayList<>();
+
+    // TODO add encoding
+    public HeadHandler(int numlines) {
+        this.numlines = numlines;
+    }
+
+    public List<String> getLines() {
+        return lines;
+    }
+
+    @Override
+    public void processStream(InputStream input) throws IOException {
+        // TODO specify encoding for InputStreamReader constructor
+        BufferedReader reader = new BufferedReader(new InputStreamReader(input));
+        int linenum = 0;
+        while (linenum < numlines) {
+            // TODO encoding
+            String line = reader.readLine();
+            if (line == null) { // EOF
+                return;
+            }
+            lines.add(line);
+            linenum++;
+        }
+        // Read and forget the rest. Need to use read() since it is not guaranteed
+        // that newline will ensue.
+        // TODO: would readline() get the job done ?
+        while (reader.read() != -1) {
```

Comment:
Technically this is doing needless `byte` to `char` conversion. E.g. were it a multi-byte encoding, then decoding is not worthwhile.

My original thinking on `HeadHandler` was to have the principal buffering at the byte-level. E.g. a `BufferedReader(InputStreamReader, /*tiny*/200)` atop a `BufferedInputStream`. Then when all the needed `readLine()` calls are done to just `read()` and discard from the `BufferedInputStream` — probably not a byte at a time but with say a very small `byte[1024]` buffer.

(Using `readLine()` here would defeat the goal of avoiding needless heap allocation.)

Reply:
I see. Now, `BufferedReader.read()` will not read a single byte from the underlying stream - it will call `fill()` to fill the internal buffer.

Reply:
Anyhow, the suggested layering sounds good so I have implemented that, together with charset choice.